Adds sprinting to humans

Incoming5643 opened the pull request:
Update 10/14

 The ` key has been made a hotkey for sprinting. Tapping it will toggle between walking and running, holding the key will enable sprinting.

Walking, running, and sprinting have been made proper defines

Sprinting across any slippable will act as if you hit lube.

---

![sprintbig](https://cloud.githubusercontent.com/assets/4176358/10440894/14402f9a-7111-11e5-9fd2-d0f2b7ad8f42.png)

(It's accessible from the UI by doubleclicking the walk/run button)

While a human sprints they ignore all slowdowns, but take stamina damage for every step they take. How much damage they take depends on how slow they would normally run. The more they're negating the more stamina damage they take. Keep in mind that since having a lot of damage also has its own slowdown, this effect gets worse if you sprint too much without taking a break to recover stamina.

If a human takes enough stamina damage (in addition to any other damage they might have) they WON'T fall into crit, but instead will get weakened(5) and have their stamina damage set to just under the threshold to where they would fall into crit.

Keep in mind this means that if you fell into stamina weakened and someone damaged you, you're combined damage pool would proc the weaken again, effectively keeping you stunlocked until you're beaten to death. This is why sprinting is so dangerous to do if you're libel to get beaten up. If you don't lose your aggressors fast, you're much more vulnerable than before.

Some benchmarks for distance. This is running down the right side hall on box station, starting from the captains office and running right:

Healthy in a uniform: Falls at the escape airlock
30 Brute uniform: Falls at Botany
60 Brute uniform: Falls at Medbay Entrance (15 tiles/1 screen)
Healthy EVA suit: Falls between Chapel and Library
30 Brute EVA suit: Falls at the Kitchen Window
60 Brute EVA suit: Falls at Medbay Entrance (15 tiles/1 screen)

Due to how life works in ticks, you can be fairly assured to be able to sprint 15 tiles straight even if you're on death's door before the stamina damage catches up to you.

(Currently humans only by nature of stamina damage not really doing anything to other carbons)

Incoming5643 committed:
Adds sprinting to humans. It's accessible from the UI by doubleclicking the walk/run button.

While a human sprints they ignore all slowdowns, but take stamina damage for every step they take. How much damage they take depends on how slow they would normally run. The more they're negating the more stamina damage they take. Keep in mind that since having a lot of damage also has its own slowdown, this effect gets worse if you sprint too much without taking a break to recover stamina.

If a human takes enough stamina damage (in addition to any other damage they might have) they WON'T fall into crit, but instead will get weakened(5) and have their stamina damage set to just under the threshold to where they would fall into crit.

Keep in mind this means that if you fell into stamina weakened and someone damaged you, you're combined damage pool would proc the weaken again, effectively keeping you stunlocked until you're beaten to death. This is why sprinting is so dangerous to do if you're in danger. If you don't lose your aggressors fast, you're much more vulnerable than before.

Some benchmarks for distance. This is running down the right side hall on box station, starting from the captains office and running right:

Healthy in a uniform: Falls at the escape airlock
30 Brute uniform: Falls at Botany
60 Brute uniform: Falls at Medbay Entrance (15 tiles/1 screen)
Healthy EVA suit: Falls between Chapel and Library
30 Brute EVA suit: Falls at the Kitchen Window
60 Brute EVA suit: Falls at Medbay Entrance (15 tiles/1 screen)

Due to how life works in ticks, you can be fairly assured to be able to sprint 15 tiles straight even if you're on death's door before the stamina damage catches up to you.

Anonus commented:
GOTTA GO FAST

Incoming5643 commented:
BUT AT WHAT COST?

Supermichael777 commented:
failed due to a lack of dive mechanic

Incoming5643 commented:
The command "python tools/ss13_genchangelog.py html/changelog.html html/changelogs" exited with 1.

I don't know what this is but it doesn't sound like my fault

Anonus commented:
This will probably be useful for shadowlings, as I believe they heal all stamina damage if they're in darkness.

Incoming5643 committed:
Forgot to close a comment in the changelog!

KorPhaeron commented:
I'd prefer the sprint just stop if you reach low enough stamina. Passing out after running down a hall seems odd.

Cool stuff though.

Incoming5643 commented:
That would require me to check human health every step they take in sprint, it sounded expensive. Also I'm kind of a masocist and the thought of overconfident spessmand trying to sprint a mile then getting dunked on sounds fun.

KorPhaeron commented:
Nothing is too expensive for speed.

And we can compromise by removing hulk or something from life checks

Zilenan91 commented:
I'm all for removing Hulk, but yeah, passing out after running 15 tiles seems a bit shitty.

RemieRichards commented:
double the 15 to 30, and this will probably be good enough.

KorPhaeron commented:
This is also probably gonna cause huge balance issues for anything not human as well. Going to be very hard for any simple animal or cyborg to catch people.

RemieRichards commented:
@KorPhaeron Do as humanity did in the past:
Humans were NOT the fastest animal, but we had one of the highest endurances.
We would literally follow our prey until they collapsed from being too tired.

GunHog commented:
Speeeeeeeeeeeeeeed.

Iamgoofball commented:
combine with meth + ephedrine for good times

Incoming5643 commented:
at current rates you can really only sprint for like four or five seconds. It's enough to rush away from some simple mob threats, but anything with a brain can follow your trail, and you'll be packing extra damage for quite a while.

You basically get one chance at this.

Zilenan91 commented:
gotta maximize your gains

WJohn commented:
Make it so lizards can't sprint for nearly as long! HUMAN STAMINA MASTER RACE.

PKPenguin321 commented:
will only be truly useful if it gets a hotkey

Incoming5643 commented:
did not know walk/run didn't have a hot key

Xhuis commented:
![](https://images.encyclopediadramatica.se/9/9f/GOTTA_GO_FAST.gif)

Dorsisdwarf commented:
@Zilenan91 the way this was described you get 15 tiles if you're at deaths door in a space suit. caps office to escape is like 50 tiles

Giacom commented:
I think this needs a hotkey, for both WASD and classic mode. You could just bind something to toggle walk or run and if they double tap then it'll register.

Giacom commented on code/modules/mob/mob_movement.dm:
```diff
@@ -157,6 +157,8 @@
 
 		switch(mob.m_intent)
+			if("sprint")
+				move_delay += config.run_speed
```
I think your sprinting speed should be affected by the damage on your legs or something. It would seem silly if you are on the edge of dying, with all the damage slowdowns and maybe a coldness slowdown, and you are able to sprint like it is nothing.

Jalleo replied:
Actually it would be better to have the damage affect your stamina instead of speed.

Incoming5643 replied:
the more damage a person has, the less they can sprint before landing on their face. If they have the damage slowdown it gets comically short

Giacom commented on code/_onclick/hud/screen_objects.dm:
```diff
@@ -145,12 +145,19 @@
 
 /obj/screen/mov_intent
-	name = "run/walk toggle"
+	name = "sprint/run/walk toggle"
```
Could you add a 'desc' on how to activate the sprint.

Giacom commented on code/game/objects/items/toys.dm:
```diff
@@ -536,7 +536,7 @@
 	if((ishuman(H))) //i guess carp and shit shouldn't set them off
 		var/mob/living/carbon/M = H
-		if(M.m_intent == "run")
+		if(M.m_intent != "walk")
```
You could do a lot of good if you changed "run" and "walk" to DEFINES with integers.

```
#define WALK 0
#define RUN 1
#define SPRINT 2
```

As I can imagine BYOND not being the fastest with comparing strings.

Supermichael777 commented:
people sliping while sprinting need to go flying

Miauw62 commented:
>people sliping while sprinting need to go flying

PLEASE

Cookiesndip commented:
Might be good, might be horrible. It would be really nice if it stopped when it would crit you because that would be really damn silly (even if it is hilarious). It's probably horrendously unbalanced right now though.

It SHOULD be short-range (speed for guaranteed death if you don't get away), and slings/thralls are going to be sprinting around in the dark making it really fucking tough to kill them since they have passive stamina regeneration. It probably shouldn't, or only partially go through damage slowdowns. I can think of nothing more frustrating than beating someone to 5 HP for them to sprint out into the hall, unable to catch them. Largely a buff to non-antags, since it's suicide for antags, esp. solo ones to ever use this function without drugs, ESPECIALLY in a chase, but the only risk for an officer is having to fall back behind the validhunter/sec assault squad or, if in a public space, retreat.

I dunno.

Xhuis commented:
Shadowlings and thralls do not have passive stamina regeneration, even in the dark.

Jalleo commented on code/_onclick/hud/human.dm:
```diff
@@ -56,7 +56,7 @@
 	using = new /obj/screen/mov_intent()
 	using.icon = ui_style
-	using.icon_state = (mymob.m_intent == "run" ? "running" : "walking")
+	using.icon_state = (mymob.m_intent == "run" ? "running" : ("walk" ? "walking" : "sprinting"))
```
Colons?

JJRcop replied:
This is a different kind of colon than the one we don't like.

duncathan replied:
Yeah, there's nothing wrong with the ternary operator.

RemieRichards replied:
dont bully the ternary

Incoming5643 replied:
#notallcolons

MrPerson replied:
Why aren't these icon states just the same as the goddamn variable to begin with.

What the FUCK

Incoming5643 commented:
So a slight problem with giving sprint/run/walk a hotkey, all the decent ones are taken. 

The closest ones left are Y and H which are a bit horrible to try and hit while holding CTRL.

` is also close but is that valid as a hot key?

JJRcop commented:
Are there hotkeys that war not used much? You can move em!

MrPerson commented:
proptip: shift

Incoming5643 commented:
You can't actually use Shift as a hotkey because byond uses it for combinations (aka you could set a hotkey for SHIFT+key just like you could CTRL+key)

MrPerson commented:
I literally just tested it and yes, you can in fact use shift alone as a macro.

Incoming5643 commented:
Ok well let me append. If you use the actual editor byond provides for that kind of file, it doesn't suggest that you can.

Anyway lot of fun ideas here I'll work on em soon.

Boredone commented:
Minor problem with just using shift.
Pretty sure that Shift Clicking people is how you quickly examine them, so that could cause problems.

Jalleo commented:
Yes yes it would

Dorsisdwarf commented:
correct me if I'm wrong

Incoming5643 commented:
I don't know if it's really controversal. Some people are saying "this might hidiously break game balance" but no one sounds like they're like, unwilling to see if it actually would.

KorPhaeron commented:
We can always use the test PR button first.

Let me know when you're ready to do so.

Incoming5643 commented:
Not yet. I want to:
*Figure out a decent hotkey
*Do some of the optimizations @Giacom mentioned (Giacom code review, nostalgic!)

Incoming5643 commented:
So how the heck do hotkeys even work in code?

As an example here's the definition of CTRL+R, which is toggle throwing

	elem 
		name = "CTRL+R"
		command = ".southwest"
		is-disabled = false

Where in the world does it actually relate to throwing?

RemieRichards commented:
@Incoming5643 ```.southwest``` is ```client/Southwest()```, which we use for toggling throwing.
Git gud.

Incoming5643 committed:
The ` key has been made a hotkey for spriting. Tapping it will toggle between walking and running, holding the key will enable sprinting.

Walking, running, and sprinting have been made proper defines

Sprinting across any slippable will act as if you hit lube.

Incoming5643 commented:
@KorPhaeron If you want to do any test rounds with this it should be ok whenever. If the test button doesn't publish the related changelog you'll have to make an annoucement for it though.

Double click the button to start sprinting. The associate hotkey is `, and you hold it to start sprinting

Incoming5643 commented:
Known bug from live testing: You can't hold ` to activate sprinting in hotkey mode while moving because wasd are technically hotkeys.

If anyone has any idea how to add doubleclick detection to a verb I'd like to hear it.

freerealestate commented:
Holding keys doesn't really work in BYOND due to lag, even if you consistently hold the key it's not consistently registered that you're doing so

RemieRichards commented on code/__DEFINES/stat.dm:
```diff
@@ -33,3 +33,8 @@
 #define SHUTTLE_ENDGAME 6
+
+//Movement modes
+#define WALK 0
```
MrPerson wanted these to match the names of the sprites
eg: 
#define WALK "walking"
#define RUN "running"
to avoid the trinary operator, plus we can do things like "\he is [M_intent]" for "he is walking" etc.

Giacom replied:
I would look into how optimized the BYOND compiler is first. An optimized compiler would see this code:

```C++
m_intent != "walking" // Would be turned from WALK into this during the preprocessor stage.
```

And it would then see that "walking" is constant and would reuse it, never recreating it more than once and never needing to calculate the hash, for equality tests.

In BYOND though? I do not know how it works. If we do assume that this is the case, we could move the \#define into a var/const value. It would be like any other global value and it should be calculated only once.

Incoming5643 replied:
I can see if this works, but I probably won't be able to practically tell the difference in terms of performance on a test server

Incoming5643 commented:
Alright so the sprinting has been testing on sybil all night and the big balance issue that's reared its head is that hulks can sprint forever because they're apparently immune to stamina damage.

As far as "making sense with the lore" it does make sense that hulks wouldn't get tired from sprinting. But from a balance perspective I don't know if I want to be the guy who buffed hulks.

Hulks either need to take normal/percentage reduced stamina damage, need to be locked out from sprinting, or need to take a different sort of damage (oxy?) while sprinting. I prefer the first one because either of the other two require a hulk check every time someone moves a tile in sprinting.

MrStonedOne commented:
for double click macros, just make the logic in the verb it triggers check for two quick presses.

note: on the second press you should put it back to where it was, so if it was at run, pressing it once would put it to walk pressing it quickly again would move it to sprint, but you should make sure to also undo the walk part, since the user made it clear they wanted to sprint, not walk.

Jalleo commented:
What about having a bit of OOP and have a proc on the human get called for when it is sprinting and have a overide proc for hulks which makes the hulk hungry and if sprinting too much puts em to sleep once they lost all their nutrition

Incoming5643 commented:
Shuttering this for now. Seeing it in action has soured the taste a bit for me and most people I've talked to were either ambivalent or negative about what seemed like a fairly needless skill ceiling raise.

Like most things I walk away from, it might come back later in a different form. Likewise if someone else has an idea to make this more decent, all the blessing in the world.
